sdks/python/apache_beam/io/bigquery.py

```diff
@@ -670,8 +701,11 @@ def _get_query_results(self, project_id, job_id,
     response = self.client.jobs.GetQueryResults(request)
     return response
 
-  @retry.with_exponential_backoff()  # Using retry defaults from utils/retry.py
+  @retry.with_exponential_backoff(
+      num_retries=MAX_RETRIES,
+      retry_filter=retry.retry_on_server_errors_and_timeout_filter)
   def _insert_all_rows(self, project_id, dataset_id, table_id, rows):
+    # TODO(SourabhBajaj): The retry might add duplicate rows here
```

Comment:
Please fix this TODO by handling timeouts here (we should not be inserting duplicate rows due to other server errors, request should fail for such errors). 

You should be able to do this by using job_id to see if a job was already inserted. If this cannot be done we should go back to not retrying on timeouts here.